feat(doc) : many improvements (#66)
aerogear/mobile-security-service: README.adoc

ifdef::env-github[]
:status:
:tip-caption: :bulb:

> :note-caption: :information_source:

Reviewer: @camilamacedo86 I noticed that you changed some of the values here. I also used these values in the UI README, should we update those also for consistency?

Author: I changed the images because they are wrong. I added the same images that are been in used in the inter8ly. Note that before this change when we use a NOTE it was shown a clip which is an attachment.

Reviewer: @camilamacedo86 Could you update them in the README for the UI as part of your follow on pr?

NOTE: This is using the `dep` package manager under the hood. You will see the dependencies installed in the `vendor` folder.

=== Configuring the Database

This REST Service is using a https://www.postgresql.org/[PostgreSQL] database. The server will not start without a https://www.postgresql.org/[PostgreSQL] available.

> In order to attend this need has two options; install the database locally or use a docker container with it. Bellow following the steps.

Reviewer 2: `Below` instead of `Bellow`

Author: Solved 👍 . Really tks.

==== By using the docker image of this project

* Removing any existing docker Postgres containers

[source,shell]
----
$ docker rm <container_id>
----

* Starting the database container

[source,shell]
----
$ docker-compose up -d db
----

NOTE: When the database is started the link:./pkg/db/db.go[db.go] routine will be executed and the database model will be created.
TIP: You can use the https://www.pgadmin.org/[pgadmin] to work with the database. See link:#installing-and-configuring-client-tool-for-the-database[Installing and configuring client tool for the database] how to setup it in your development environment.

* Checking the database

> Following the commands to access the database in the container and check if the tables were created with success.

Reviewer 2: `successfully` rather than `with success`

Author: Solved 👍 . Really tks.

NOTE: This is using the `dep` package manager under the hood. You will see the dependencies installed in the `vendor` folder.

=== Configuring the Database

> This REST Service is using a https://www.postgresql.org/[PostgreSQL] database. The server will not start without a https://www.postgresql.org/[PostgreSQL] available.

Reviewer 3: .Prerequisite

* A https://www.postgresql.org/[PostgreSQL] database

Reviewer 3: but if this is a requirment, why are we telling them how to set it up?

> In order to attend this need has two options; install the database locally or use a docker container with it. Below following the steps.

Reviewer: ```suggestion
There are two ways to install the database:
```

==== By using the docker image of this project

* Removing any existing docker Postgres containers

[source,shell]
----
$ docker rm <container_id>
----

* Starting the database container

[source,shell]
----
$ docker-compose up -d db
----

NOTE: When the database is started the link:./pkg/db/db.go[db.go] routine will be executed and the database model will be created.

> TIP: You can use the https://www.pgadmin.org/[pgadmin] to work with the database. See link:#installing-and-configuring-client-tool-for-the-database[Installing and configuring client tool for the database] how to setup it in your development environment.

Reviewer: ```suggestion
TIP: You can use https://www.pgadmin.org/[pgadmin] to work with the database. See link:#installing-and-configuring-client-tool-for-the-database[Installing and configuring client tool for the database] 
```

* Checking the database

> Following the commands to access the database in the container and check if the tables were created successfully.

Reviewer: I think we can remove this line. Each step below already details what it's doing.

. Run `$ docker ps` to get the `container id`
. Run `$docker exec -it <postgres-container-id> bash` to access the container via shell
. Run `psql -U postgres` to login with the user `postgres` in the database
. Run `\c mobile_security_service` to connect in the database created for this project
. Run `\dt` to lists all tables in the current database

==== By using a local Postgre DB

The default values to connect in the database can be found in the file link:./pkg/config/config.go[config.go].

> . Setup your local database and/or change the configuration defined.

Reviewer: suggestion```
Set up your local database and/or change the configuration defined.
```

. Create the database and tables in your local PostgreSQL installation. See the file link:./pkg/db/db.go[db.go] to get the scripts to create the tables.

> NOTE: The link:./pkg/config/config.go[config.go] will use the default values, however, if would like to change them it should be made by link:#environment-variables[Environment Variables], see the topic link:#adding-your-own-.env-file[Adding your own .env file]

Reviewer: suggestion```
NOTE: The file link:./pkg/config/config.go[config.go] will use the default values, however, if would like to change them it should be made by link:#environment-variables[Environment Variables], see the topic link:#adding-your-own-.env-file[Adding your own .env file]
```

Reviewer: OR
```suggestion
NOTE: link:./pkg/config/config.go[config.go] will use the default values, however, if would like to change them it should be made by link:#environment-variables[Environment Variables], see the topic link:#adding-your-own-.env-file[Adding your own .env file]
```

=== Starting the Server

[source,shell]
----
go run cmd/mobile-security-service/main.go
----

=== Checking REST Service

> Following a curl call to check if the service is running as should be.

Reviewer: ```suggestion
Run the following curl command and check the output is as shown
```

=== Building & Testing

The `Makefile` provides commands for building and testing the code. Some dependencies are required to run these commands.

==== Installing the required dependencies

Dependencies may be required to run some of the `Make` commands. Below are instructions on how to install them.

Run the following command.

[source,shell]
----
$ go get -u github.com/matryer/moq
----

NOTE : See all commands available in link:#using-make-commands[Using make commands]

=== Installing and configuring client tool for the database

You can use the https://www.pgadmin.org/[pgadmin] which is the client tool for PostgreSQL to work with the database.

* Download and install the client tool

NOTE: The link to download for MacOS is : https://www.pgadmin.org/download/pgadmin-4-macos/

* Configure the client tool

Following the steps to do this setup.

. Access the tool via the browser. The default link will be `http://127.0.0.1:52263/browser/#`
. Create a new server connection with the database. Following the image to show how to do it.

image::https://user-images.githubusercontent.com/7708031/53171792-9ecf3380-35db-11e9-8de7-4a7df979b38b.png[Create Server,align="center"]

> . Add the data required to connect on the database. See the values defined in link:./pkg/config/config.go[config.go]. Following an image with this setup.

Reviewer 2: `connect to the database` rather that `connect on the database`

Author: Solved 👍 . Really tks.

image::https://user-images.githubusercontent.com/7708031/53172136-857ab700-35dc-11e9-9794-4fa016703135.png[Configuration,640,align="center"]

=== Adding data to test the Service

Use the scripts from the file link:./pkg/test/fixtures.go[fixtures.go]. It will allow you have data to check this service. 

=== Creating mock files for the interfaces

> This project is using teh dependency https://github.com/matryer/moq[moq]. Follow the steps below to use it.

Reviewer 2: the

Author: 👍 great

The Swagger UI is available at http://localhost:8080[localhost:8080].

> ==== By browser

Reviewer 3: this seems easier, should it be the first option?

There is also a https://chrome.google.com/webstore/detail/swagger-ui-console/ljlmonadebogfjabhkppkoohjkjclfai?hl=en[Chrome extension] you can use instead of running a Docker container.

Paste https://raw.githubusercontent.com/aerogear/mobile-security-service/master/api/swagger.yaml[https://raw.githubusercontent.com/aerogear/mobile-security-service/master/api/swagger.yaml] and press **Explore**.

> TIP: Also, you can use the https://petstore.swagger.io/[Demo] tool online for swaggerUI. 

Reviewer 3: maybe this should be the first option ;)

[source,shell]
----
$ go get -u github.com/matryer/moq
----

NOTE : See all commands available in link:#using-make-commands[Using make commands]

=== Installing and configuring client tool for the database

You can use the https://www.pgadmin.org/[pgadmin] which is the client tool for PostgreSQL to work with the database.

* Download and install the client tool

NOTE: The link to download for MacOS is : https://www.pgadmin.org/download/pgadmin-4-macos/

* Configure the client tool

Following the steps to do this setup.

. Access the tool via the browser. The default link will be `http://127.0.0.1:52263/browser/#`
. Create a new server connection with the database. Following the image to show how to do it.

image::https://user-images.githubusercontent.com/7708031/53171792-9ecf3380-35db-11e9-8de7-4a7df979b38b.png[Create Server,align="center"]

. Add the data required to connect to the database. See the values defined in link:./pkg/config/config.go[config.go]. Following an image with this setup.

image::https://user-images.githubusercontent.com/7708031/53172136-857ab700-35dc-11e9-9794-4fa016703135.png[Configuration,640,align="center"]

=== Adding data to test the Service

> Use the scripts from the file link:./pkg/test/fixtures.go[fixtures.go]. It will allow you have data to check this service. 

Reviewer 3: +1

Author: What do you mean here with +1?

Reviewer 3: just like the link to data script ;)

NOTE: See more over it in the Readme of https://github.com/matryer/moq[moq]

* Using the mock

> In the created file you will see an implementation commented as the following example. This implementation will be used in the test file to mock the methods/func of this interface.

Reviewer 2: `as in the following example` rather than `as the following example`

Author: Solved 👍 . Really tks.

[source,go]
----
// AppServiceMock is a mock implementation of Service.
//
//     func TestSomethingThatUsesAppService(t *testing.T) {
//
//         // make and configure a mocked Service
//         mockedAppService := &AppServiceMock{
//             GetAppsFunc: func() (*[]models.App, error) {
// 	               panic("mock out the GetApps method")
//             },
//         }
//
//         // use mockedAppService in code that requires Service
//         // and then make assertions.
//
//     }
----

* Adding data to the mock function

> The `panic` statement need to be replaced for the mock data. Following an example.

Reviewer 2: `needs` rather that `need`

Author: Solved 👍 . Really tks.

NOTE: See more over it in the Readme of https://github.com/matryer/moq[moq]

* Using the mock

In the created file you will see an implementation commented as in the following example. This implementation will be used in the test file to mock the methods/func of this interface.

[source,go]
----
// AppServiceMock is a mock implementation of Service.
//
//     func TestSomethingThatUsesAppService(t *testing.T) {
//
//         // make and configure a mocked Service
//         mockedAppService := &AppServiceMock{
//             GetAppsFunc: func() (*[]models.App, error) {
// 	               panic("mock out the GetApps method")
//             },
//         }
//
//         // use mockedAppService in code that requires Service
//         // and then make assertions.
//
//     }
----

> * Adding data to the mock function

Reviewer 3: the word 'mock' works well here, but in other headings it's not so good, eg 'Editing the mock'. 
@austincunningham  can we avoid using an adjective as a noun, even if we use 'mock system' or similar?

The `panic` statement needs to be replaced for the mock data. Following an example.

[source,go]
----
numOfDeployedVersions := 5
numOfAppLaunches := 1000
numOfCurrentInstalls := 9000

// mock data
app := models.App{
    ID:                    "a0874c82-2b7f-11e9-b210-d663bd873d93",
    AppID:                 "com.aerogear.app1",
    AppName:               "app1",
    NumOfDeployedVersions: &numOfDeployedVersions,
    NumOfAppLaunches:      &numOfAppLaunches,
    NumOfCurrentInstalls:  &numOfCurrentInstalls,
}

// make and configure a mocked Service
mockedAppService := &AppServiceMock{
    GetAppsFunc: func() (*[]models.App, error) {
        return &[]models.App{
            app,
        }, nil
    },
}
----

* Calling the mock

You will call the mock instead of use the interface. It will return the data mocked as defined above. Following an practical example.

[source,go]
----
got, err := mockedAppService.GetApps()
if (err != nil) != tt.wantErr {
    t.Errorf("appsService.GetApps() error = %v, wantErr %v", err, tt.wantErr)
    return
}
if !reflect.DeepEqual(got, tt.want) {
    t.Errorf("appsService.GetApps() = %v, want %v", got, tt.want)
}
----

> NOTE: The mock file contains explanations for example over how to check the quantity of results.

Reviewer: @camilamacedo86 I'm not sure what you are trying to say here.
Is it? 
```suggestion
NOTE: The mock file contains explanations for how to check the quantity of results.
```

Reviewer: @camilamacedo86 This one is still outstanding.

Author: Sure, you are 100% correct here. It is fixed now. Tks to get it.

| `make docker_push_master`     | Pushes master image to Docker image hosting repository
|===

=== Using Postman to test the REST API

> You can install the https://www.getpostman.com/[Postman] tool which will be useful to call and test the REST API endpoints of this server.

Reviewer: ```suggestion
TIP: You can install the https://www.getpostman.com/[Postman] tool which will be useful to call and test the REST API endpoints of this server.
```

Reviewer: The reason I suggest this change is because I agree with @craicoverflow's comments above. I think it's fine to suggest it as a tip but it's not part of the project so is not necessary for the user to use this tool to run the project.

Reviewer: Also after reviewing the file in full I do think that we should remove the image. It could get outdated quickly and it's not a tool that we are providing. The link out to postman should be sufficient.

Author: removed the topic and added as a TIP where has the curl requested.

> Following an image to show how it works.

Reviewer: I'm not sure we need this line. 
I think it can either be removed or added to the above line. Something like 
..... endpoints of this server (See image below).

Reviewer: As per my comment above, I suggest removing the image.

Author: It was removed

image::https://user-images.githubusercontent.com/7708031/53174794-8878a600-35e2-11e9-889d-79973d9e16db.png[Postman]

> == Built With

Reviewer 3: @camilamacedo86  would this make more sense as a badge at the top of the page?

Reviewer 4: @finp @camilamacedo86 I'm not sure if Postman should be in the README. It's an external tool for making REST calls but it is nothing to do with this project. Generally, if somebody wants to work on a project like this they would know what Postman is and they may prefer to use something else (their IDE, CURL etc). I don't think a screenshot of Postman should be in here showing them how to use it - it is pretty straightforward anyway and it only serves to make our documentation more busy than it should be. What do you think?

Author: @craicoverflow the idea here is to have in the documented how a dev can contribute with and some tips.
